### DISCUSSION TITLE
Support attribute downloading

### DISCUSSION THREAD

**Discussion by qgustavor** [qgustavor/mega#96] [Ideas]
Currently only uploading attributes (thumbnail and preview images) are supported.

**Comment by qgustavor** [comment]
Attribute handlers are stored on the `.fa` property returned by the server when listing files in this format: `123:0*aaaaaaaaaaa/456:1*bbbbbbbbbbb`. It's a string represeting each attribute separated with `/`. The attribute with `0*` is the thumbnail and `1*` is preview image. I don't know what `123` and `456` represents. 

Encrypted attributes can be downloaded by requesting their upload URL using `{a: "ufa", fah: "attribute handler", ssl: 2, r: 1}`. SSL parameter works similar to the one used when downloading files (setting it to 0 returns a HTTP URL, setting it to 2 returns a HTTPS URL). The URL is returned in `.p` property. The returned contents need to be decrypted using AES-CBC using the file decryption key.

Seems `.fa` is also used to stored video codec information on the `*8` attribute. Stored info is FPS, width, height, audio codec name, video codec name, video duration, video container and "shortformat". Would be a good thing implementing parsing and updating this attribute?

**Edit:** video informations are stored encrypted with [XXTEA](https://en.wikipedia.org/wiki/XXTEA). Why not AES? Maybe they wanted a smaller block size. Why not XSalsa20 (as MEGA already uses TweetNaCl)? I don't know...

**Comment by qgustavor** [comment]
> Why not AES? Maybe they wanted a smaller block size. Why not XSalsa20 (as MEGA already uses TweetNaCl)? I don't know...

After working [on this](https://github.com/qgustavor/generic-speck) I know the answer: because XSalsa20 is not a block cipher, is a stream cipher.

Looks like they want to reduce the space used by the ciphertext, then they can't use a stream cipher because it would require to store a nonce and replace it every time otherwise information would leak. The same issue don't happens with block ciphers.

**Comment by AlttiRi** [comment]
`123` and `456` are bunch (group) number for the bulk downloading.
The thumbs/prevs with same bunch number can be requested within one download request.

You can debug a demo (node, or browser) of my unfinished library https://github.com/AlttiRi/meganz-api to investigate how it works.

**Comment by AlttiRi** [comment]
Here are my old first demo tests of that:
https://github.com/AlttiRi/meganz-api/blob/master/__experiments/app/app-test-file-attributes.js
https://github.com/AlttiRi/meganz-api/blob/master/__experiments/app/file-attributes-old.js

The current code in `/src/` is more complicated:
https://github.com/AlttiRi/meganz-api/blob/master/src/file-attributes.js
